### ISSUE TITLE
createStone.solo throws a UndefinedObject does not understand  #'asFileReference'

### ISSUE THREAD

**Issue by JupiterJones** [GsDevKit/GsDevKit_stones#2]
I'm probably just doing something wrong.

```
jupiter@puppy ~ % registryReport.solo --registry=puppy
GDKStonesRegistry {
	#name : 'puppy',
	#parentRegistryName : 'puppy',
	#parentRegistryPath : '$XDG_DATA_HOME/gsdevkit_stones/registry.ston',
	#stones : { },
	#sessions : { },
	#products : {
		'3.6.5' : '/Users/jupiter/Documents/Development/GemStone/GsDevKit/GemStoneProduct/GemStone64Bit3.6.5-i386.Darwin'
	},
	#projectSets : { },
	#templates : {
		'default_seaside' : '$XDG_DATA_HOME/gsdevkit_stones/templates/default_seaside.ston',
		'minimal_rowan' : '$XDG_DATA_HOME/gsdevkit_stones/templates/minimal_rowan.ston',
		'minimal' : '$XDG_DATA_HOME/gsdevkit_stones/templates/minimal.ston',
		'minimal_seaside' : '$XDG_DATA_HOME/gsdevkit_stones/templates/minimal_seaside.ston',
		'default' : '$XDG_DATA_HOME/gsdevkit_stones/templates/default.ston',
		'default_rowan' : '$XDG_DATA_HOME/gsdevkit_stones/templates/default_rowan.ston'
	}
}
jupiter@puppy ~ % createStone.solo --root=/Users/jupiter/Documents/Development/GemStone/GsDevKit/myStones/test1 --projectsHome=/Users/jupiter/Documents/Development/GemStone/GsDevKit/myStones/git --template=minimal --registry=puppy --force test1 3.6.5
a MessageNotUnderstood occurred (error 2010), a UndefinedObject does not understand  #'asFileReference'
```

All the paths I've used in the command exist, and the 'test1' directory is created once the command starts, however it's an empty directory when the error is thrown.

**Comment by dalehenrich** [comment]
@JupiterJones if you use the -D option on the command you can get a topaz stack and that would help understand the error (which version of GemStone you are useing would be useful, too) ...

**Comment by JupiterJones** [comment]
Doesn't look that helpful :)

GemStone 3.6.5 Darwin Intel

```
jupiter@puppy ~ % createStone.solo --root=/Users/jupiter/Documents/Development/GemStone/GsDevKit/myStones/test1 --projectsHome=/Users/jupiter/Documents/Development/GemStone/GsDevKit/myStones/git --template=minimal --registry=puppy --force -D test1 3.6.5
-----------------------------------------------------
 Near line 81 of file /Users/jupiter/Documents/Development/GemStone/GsDevKit/GsDevKit_stones/superDoit/gemstone/superdoit.tpz
GemStone: Error         Nonfatal
a MessageNotUnderstood occurred (error 2010), a UndefinedObject does not understand  #'asFileReference'
Error Category: 231169 [GemStone] Number: 2010  Arg Count: 4 Context : 105150209 exception : 104941313
Arg 1: [1093121 size:15 primitiveSize:15 cls: 110849 Symbol] asFileReference
Arg 2: [2 size:0 primitiveSize:0 cls: 74241 SmallInteger] 0 == 0x0
Arg 3: [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
Arg 4: [104940801 size:0 primitiveSize:0 cls: 66817 Array] anArray
Stopping at line 81 of /Users/jupiter/Documents/Development/GemStone/GsDevKit/GsDevKit_stones/superDoit/gemstone/superdoit.tpz
```

**Comment by dalehenrich** [comment]
when you are in topaz, use the `where` command to get a full stack ... as an FYI, the `frame` command can be used to view the variables in a particular frame on the stack and if you do `display classoops` you will be able to see the oops of the variables and then `obj2 OOP` to inspect 2 levels of a particular object ... I forget that folks aren't normally familiar with the topaz command line commands ...

**Comment by JupiterJones** [comment]
Thanks for that. I clearly wasn't paying attention as I didn't even notice that I was left in topaz :)

This look as little more suspicious...

```
topaz 1> where
1 MessageNotUnderstood >> defaultAction         @2 line 3   [methId 6397697]
2 MessageNotUnderstood (AbstractException) >> _signalWith: @5 line 25   [methId 3495937]
3 MessageNotUnderstood (AbstractException) >> signal @2 line 47   [methId 3508225]
4 UndefinedObject (Object) >> doesNotUnderstand: @9 line 10   [methId 6299649]
5 UndefinedObject (Object) >> _doesNotUnderstand:args:envId:reason: @8 line 14   [methId 6255105]
6 GDKStonesRegistry >> downloadProductName:productFile:to: @2 line 5   [methId 104920321]
7 GDKStonesRegistry >> downloadProductNamed:    @14 line 8   [methId 104918785]
8 [] in GDKStonesRegistry >> productNamed:      @7 line 8   [methId 104917505]
9 Dictionary >> at:ifAbsent:                    @7 line 10   [methId 12591617]
10 GDKStonesRegistry >> productNamed:ifAbsent:   @3 line 4   [methId 104915201]
11 GDKStonesRegistry >> productNamed:            @2 line 5   [methId 104913665]
==> 12 GDKStoneSpec >> createStoneStructureUsing:    @5 line 5   [methId 104860929]
13 Executed Code                                 @50 line 26   [methId 104858881]
14 GsNMethod >> _executeInContext:               @1 line 9   [methId 5602817]
15 String >> evaluateInContext:symbolList:       @3 line 10   [methId 6033409]
16 SuperDoitExecutionClass (SuperDoitExecutionMetadataClass) >> theDoit @4 line 35   [methId 104855041]
17 [] in SuperDoitExecution >> doit              @14 line 7   [methId 104890625]
18 ExecBlock0 (ExecBlock) >> on:do:              @3 line 44   [methId 5736705]
19 SuperDoitExecutionClass (SuperDoitExecution) >> doit @4 line 8   [methId 104888065]
20 SuperDoitDoitCommand >> executeAgainst:       @27 line 19   [methId 104886785]
21 SuperDoitDoitCommand >> executeAgainst:onErrorDo: @2 line 4   [methId 104885505]
22 [] in SuperDoitCommandDefinition >> executeAgainst:onErrorDo: @7 line 3   [methId 104883713]
23 OrderedCollection (Collection) >> do:         @5 line 10   [methId 5872641]
24 SuperDoitCommandDefinition >> executeAgainst:onErrorDo: @3 line 3   [methId 104853249]
25 [] in SuperDoitCommandParser >> parseAndExecuteScriptFile: @22 line 17   [methId 104851969]
26 ExecBlock0 (ExecBlock) >> ensure:             @2 line 12   [methId 5748737]
27 SuperDoitCommandParser >> parseAndExecuteScriptFile: @4 line 53   [methId 104901121]
28 SuperDoitCommandParser class >> processInputFile @29 line 21   [methId 104899585]
29 Executed Code                                 @2 line 6   [methId 104897537]
30 GsNMethod class >> _gsReturnToC               @1 line 11   [methId 5568257]
  [GsProcess 105150209]
topaz 1> frame 7
7 GDKStonesRegistry >> downloadProductNamed:    @14 line 8   [methId 104918785]
    receiver [104893953 size:0 primitiveSize:9 cls: 105001985 GDKStonesRegistry] aGDKStonesRegistry
    productName [104913153 size:5 primitiveSize:5 cls: 74753 String] test1
    gemVersionReport [104831489 size:14 primitiveSize:24 cls: 79873 StringKeyValueDictionary] aStringKeyValueDictionary
    downloadFileName [104830721 size:32 primitiveSize:32 cls: 74753 String] GemStone64Bittest1-x86_64.Darwin
    productDirPath [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
    .t1 [104893953 size:0 primitiveSize:9 cls: 105001985 GDKStonesRegistry] aGDKStonesRegistry
    .t2 [104913153 size:5 primitiveSize:5 cls: 74753 String] test1
    .t3 [104830721 size:32 primitiveSize:32 cls: 74753 String] GemStone64Bittest1-x86_64.Darwin
    .t4 [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
topaz 1> frame 6
6 GDKStonesRegistry >> downloadProductName:productFile:to: @2 line 5   [methId 104920321]
    receiver [104893953 size:0 primitiveSize:9 cls: 105001985 GDKStonesRegistry] aGDKStonesRegistry
    productName [104913153 size:5 primitiveSize:5 cls: 74753 String] test1
    downloadFileName [104830721 size:32 primitiveSize:32 cls: 74753 String] GemStone64Bittest1-x86_64.Darwin
    downloadPath [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
    cwd [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
    downloadDir [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
    .t1 [104828161 size:6 primitiveSize:9 cls: 134913 VariableContext] aVariableContext
    .t2 [20 size:0 primitiveSize:0 cls: 76289 UndefinedObject] nil
    .t3 [1093121 size:15 primitiveSize:15 cls: 110849 Symbol] asFileReference
    .t4 [104940801 size:0 primitiveSize:0 cls: 66817 Array] anArray
    .t5 [2 size:0 primitiveSize:0 cls: 74241 SmallInteger] 0 == 0x0
    .t6 [2 size:0 primitiveSize:0 cls: 74241 SmallInteger] 0 == 0x0
    _VC [104828161 size:6 primitiveSize:9 cls: 134913 VariableContext] aVariableContext
```
downloadFileName is 'GemStone64Bittest1-x86_64.Darwin' - looks like the stoneName has been added it to.
productPath is nil

I downloaded 3.6.5 manually and added it to the registry (registerProduct.solo) before running this command, so I'm not sure why it is trying to download it.

Maybe I don't understand how my directory structure should be setup by default.

**Comment by dalehenrich** [comment]
this looks like enough information for me to figure out what is going on ... Thanks!!!

**Comment by dalehenrich** [comment]
@JupiterJones, I'm sorry is this a blocking issue? If it is I've got my head far enough above water, where I can address this if needed ...

**Comment by JupiterJones** [comment]
Not blocking at all. No hurry :)

**Comment by feldti** [comment]
Same problem here with:  downloadGemStone.solo --registry=$registryName 3.6.6

==> 1 MessageNotUnderstood >> defaultAction         @9 line 7
2 MessageNotUnderstood (AbstractException) >> _defaultAction @4 line 4
3 MessageNotUnderstood (AbstractException) >> _signal @2 line 20
4 MessageNotUnderstood (AbstractException) >> signal @2 line 47
5 UndefinedObject (Object) >> doesNotUnderstand: @9 line 10
6 UndefinedObject (Object) >> _doesNotUnderstand:args:envId:reason: @12 line 17
7 GDKStonesRegistry class >> downloadProductName:productDirectory:to: @7 line 7
8 GDKStonesRegistry >> downloadProductName:productDirectory:to: @3 line 3
9 GDKStonesRegistry >> downloadProductNamed:    @7 line 6
10 [] in GDKStonesRegistry >> productNamed:      @7 line 8
11 Dictionary >> at:ifAbsent:                    @7 line 10
12 GDKStonesRegistry >> productNamed:ifAbsent:   @3 line 4
13 GDKStonesRegistry >> productNamed:            @2 line 5
14 SuperDoitExecutionClass >> downloadToRegistry:version:productName: @8 line 6
15 Executed Code                                 @44 line 23
16 GsNMethod >> _executeInContext:               @1 line 11
17 String >> evaluateInContext:symbolList:       @3 line 10
18 SuperDoitExecutionClass (SuperDoitExecutionMetadataClass) >> theDoit @4 line 32
19 [] in SuperDoitExecution >> doit              @20 line 6
20 ExecBlock0 (ExecBlock) >> on:do:              @3 line 44
21 SuperDoitExecutionClass (SuperDoitExecution) >> doit @4 line 8
22 SuperDoitDoitCommand >> executeAgainst:       @27 line 19
23 SuperDoitDoitCommand >> executeAgainst:onErrorDo: @2 line 4
24 [] in SuperDoitCommandDefinition >> executeAgainst:onErrorDo: @7 line 3
25 OrderedCollection (Collection) >> do:         @5 line 10
26 SuperDoitCommandDefinition >> executeAgainst:onErrorDo: @3 line 3
27 [] in SuperDoitCommandParser >> parseAndExecuteScriptFile: @22 line 17
28 ExecBlock0 (ExecBlock) >> ensure:             @2 line 12
29 SuperDoitCommandParser >> parseAndExecuteScriptFile: @4 line 53
30 SuperDoitCommandParser class >> processInputFile @29 line 21
31 Executed Code                                 @2 line 6
32 GsNMethod class >> _gsReturnToC               @1 line 11

**Comment by feldti** [comment]
Ok, I understand, that the scripts do not check that all needed variables are set in the registry

**Comment by dalehenrich** [comment]
Which branch of GsDevKit_stones are you on ... a SHA would be good in the future ... I'm looking at the source for v2 [cde8aac] and see that the productPath iv in the GDKStonesRegistry must be nil?

I assume this code is the samve in v2.1 as well ... 

I plan on making a fix for this in v2 and v2.1 ... I plan to add nil checks for all of the registry variables ...
